### ISSUE TITLE
CPP parse error with cpp 4.8.1.

### ISSUE THREAD

**Issue by master-q** [ajhc/ajhc#24]
The bug is not reproduced with cpp 4.7.3.

```
$ cpp --version
cpp (Debian 4.8.1-3) 4.8.1
$ make install
-- snip --
ajhc -L- -L. --build-hl lib/jhc/jhc.yaml -o jhc-1.0.hl
ajhc 0.8.0.6 (646a37c25c04042e543be888c82e51bf84e1b80e)
Creating library from description file: "lib/jhc/jhc.yaml"
Reading: "lib/jhc/jhc.yaml"
Finding Dependencies...
Using Ho Cache: '/home/kiwamu/.ajhc/cache'
Data.Ratio              [lib/jhc/Data/Ratio.hs]
Jhc.Basics              [lib/jhc/Jhc/Basics.hs]
Jhc.Int                 [lib/jhc/Jhc/Int.hs]
Jhc.Type.Word           [lib/jhc/Jhc/Type/Word.hs]
Jhc.Type.Basic          [lib/jhc/Jhc/Type/Basic.hs]
Jhc.Class.Real          [lib/jhc/Jhc/Class/Real.hs]
Jhc.Class.Num           [lib/jhc/Jhc/Class/Num.hs]
Jhc.Class.Ord           [lib/jhc/Jhc/Class/Ord.hs]
Jhc.Show                [lib/jhc/Jhc/Show.hs]
Writing stuff:/tmp/jhc_29537/prelude.m4
Jhc.Enum                [lib/jhc/Jhc/Enum.hs]
Jhc.Inst.PrimEnum       [lib/jhc/Jhc/Inst/PrimEnum.hs]
Jhc.Float               [lib/jhc/Jhc/Float.hs]
Jhc.Inst.Num            [lib/jhc/Jhc/Inst/Num.hs]
Jhc.Inst.Enum           [lib/jhc/Jhc/Inst/Enum.hs]
Jhc.IO                  [lib/jhc/Jhc/IO.hs]
/usr/include/stdc-predef.h:1:1: Parse error hoge
make[2]: *** [jhc-1.0.hl] エラー 1
make[2]: ディレクトリ `/home/kiwamu/src/ajhc' から出ます
```

**Comment by master-q** [comment]
Need better options for cpp command?

``` haskell
-- FIle: src/Ho/ReadSource.hs
preprocess :: Opt -> FilePath -> LBS.ByteString -> IO LBS.ByteString
preprocess opt fn lbs = do
    let fopts s = s `Set.member` optFOptsSet opt
        incFlags = [ "-I" ++ d | d <- optIncdirs opt ++ optIncs opt]
        defFlags = ("-D__JHC__=" ++ revision):("-D__JHC_VERSION__=" ++ version):[ "-D" ++ d | d <- optDefs opt]
    case () of
        _ | fopts FO.Cpp -> readSystem "cpp" $ ["-CC","-traditional"] ++ incFlags ++ defFlags ++ [fn]
          | fopts FO.M4  -> do
                m4p <- m4Prelude
                readSystem "m4" $ ["-s", "-P"] ++ incFlags ++ defFlags ++ [m4p,fn]
          | otherwise -> return lbs
```

**Comment by master-q** [comment]
```
$ cpp-4.7 -CC -traditional ./lib/jhc/Jhc/Options.hs
# 1 "./lib/jhc/Jhc/Options.hs"
# 1 "<command-line>"
# 1 "./lib/jhc/Jhc/Options.hs"
{-# OPTIONS_JHC -fno-prelude -fffi -fcpp -funboxed-values #-}
{-# LANGUAGE CPP #-}

module Jhc.Options(







    Target(..)
    ) where

import Jhc.Basics
import Jhc.Enum
import Jhc.Order

data Target = Grin | GhcHs | DotNet | Java
    deriving(Eq,Ord,Enum)
```

```
$ cpp-4.8 -CC -traditional ./lib/jhc/Jhc/Options.hs
# 1 "./lib/jhc/Jhc/Options.hs"
# 1 "<command-line>"
# 1 "/usr/include/stdc-predef.h" 1 3 4
/* Copyright (C) 1991-2012 Free Software Foundation, Inc.
   This file is part of the GNU C Library.

   The GNU C Library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.

   The GNU C Library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public
   License along with the GNU C Library; if not, see
   <http://www.gnu.org/licenses/>.  */




/* This header is separate from features.h so that the compiler can
   include it implicitly at the start of every compilation.  It must
   not itself include <features.h> or any other header that includes
   <features.h> because the implicit include comes before any feature
   test macros that may be defined in a source file before it first
   explicitly includes a system header.  GCC knows the name of this
   header in order to preinclude it.  */

/* Define __STDC_IEC_559__ and other similar macros.  */

# 1 "/usr/include/x86_64-linux-gnu/bits/predefs.h" 1 3 4
/* Copyright (C) 2005 Free Software Foundation, Inc.
   This file is part of the GNU C Library.

   The GNU C Library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.

   The GNU C Library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public
   License along with the GNU C Library; if not, write to the Free
   Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
   02111-1307 USA.  */








/* We do support the IEC 559 math functionality, real and complex.  */



# 31 "/usr/include/stdc-predef.h" 2 3 4

/* wchar_t uses ISO/IEC 10646 (2nd ed., published 2011-03-15) /
   Unicode 6.0.  */


/* We do not support C11 <threads.h>.  */


# 1 "<command-line>" 2
# 1 "./lib/jhc/Jhc/Options.hs"
{-# OPTIONS_JHC -fno-prelude -fffi -fcpp -funboxed-values #-}
{-# LANGUAGE CPP #-}

module Jhc.Options(







    Target(..)
    ) where

import Jhc.Basics
import Jhc.Enum
import Jhc.Order

data Target = Grin | GhcHs | DotNet | Java
    deriving(Eq,Ord,Enum)
```

**Comment by master-q** [comment]
GHC does not this BUG, because use own cpphs command instead of GCC's cpp.

**Comment by master-q** [comment]
Fix at c79980df7f776fb4c0494d05732b8e707f1c1194 .
